main/acle.md

```diff
@@ -9546,25 +9680,2141 @@ possible to write these operations using normal C arithmetic. For example:
+  //   _za64[_u64] (only if __ARM_FEATURE_SME_I16I64 != 0)
+  __attribute__((arm_streaming, arm_shared_za))
+  void svadd_za32[_f32]_vg1x4(uint32_t slice, svfloat32x4_t zm);
+  
+  
+  // Variants are available for:
+  //   _za32[_f32]
+  //   _za32[_u32]
+  //   _za64[_f64] (only if __ARM_FEATURE_SME_F64F64 != 0)
+  //   _za64[_u64] (only if __ARM_FEATURE_SME_I16I64 != 0)
+  __attribute__((arm_streaming, arm_shared_za))
+  void svsub_za32[_f32]_vg1x2(uint32_t slice, svfloat32x2_t zm);
+  
+  
+  // Variants are available for:
+  //   _za32[_f32]
+  //   _za32[_u32]
+  //   _za64[_f64] (only if __ARM_FEATURE_SME_F64F64 != 0)
+  //   _za64[_u64] (only if __ARM_FEATURE_SME_I16I64 != 0)
+  __attribute__((arm_streaming, arm_shared_za))
+  void svsub_za32[_f32]_vg1x4(uint32_t slice, svfloat32x4_t zm);
+  
+  ```
+
+#### BFCVTN, FCVTN
+
+Multi-vector floating-point convert from single-precision to interleaved half-precision/BFloat16
+
+``` c
+  // Variants are also available for _f16[_f32_x2]
+  __attribute__((arm_streaming))
+  svbfloat16_t svcvtn_bf16[_f32_x2](svfloat32x2_t zn);
+  
+  ```
+
+#### FCVT, BFCVT, FCVTZS, FCVTZU, SCVTF, UCVTF
+
+Multi-vector convert to/from floating-point.
+
+``` c
+  // Variants are also available for _f16[_f32_x2]
+  __attribute__((arm_streaming))
+  svbfloat16_t svcvt_bf16[_f32_x2](svfloat32x2_t zn);
+
+
+  // Variants are also available for _f32[_u32_x2], _s32[_f32_x2] and _u32[_f32_x2]
+  __attribute__((arm_streaming))
+  svfloat32x2_t svcvt_f32[_s32_x2](svint32x2_t zn);
+  
+  
+  // Variants are also available for _f32[_u32_x4], _s32[_f32_x4] and _u32[_f32_x4]
+  __attribute__((arm_streaming))
+  svfloat32x4_t svcvt_f32[_s32_x4](svint32x4_t zn);
+  
+  ```
+
+#### SQCVT, SQCVTU, UQCVT
+
+Multi-vector saturating extract narrow
+
+``` c
+  // Variants are also available for _u16[_s32_x2] and _u16[_u32_x2]
+  __attribute__((arm_streaming))
+  svint16_t svqcvt_s16[_s32_x2](svint32x2_t zn);
+  
+  
+  // Variants are also available for _u8[_s32_x4], _u8[_u32_x4], _s16[_s64_x4],
+  // _u16[_s64_x4] and _u16[_u64_x4]
+  __attribute__((arm_streaming))
+  svint8_t svqcvt_s8[_s32_x4](svint32x4_t zn);
+  
+  ```
+
+#### SQCVTN, SQCVTUN, UQCVTN
+
+Multi-vector saturating extract narrow and interleave
+
+``` c
+  // Variants are also available for _u16[_s32_x2] and _u16[_u32_x2]
+  __attribute__((arm_streaming_compatible))
+  svint16_t svqcvtn_s16[_s32_x2](svint32x2_t zn);
+  
+  
+  // Variants are also available for _u8[_s32_x4], _u8[_u32_x4], _s16[_s64_x4],
+  // _u16[_s64_x4] and _u16[_u64_x4]
+  __attribute__((arm_streaming))
+  svint8_t svqcvtn_s8[_s32_x4](svint32x4_t zn);
+  
+  ```
+
+#### UDOT, SDOT, FDOT (vectors)
+
+Multi-vector dot-product (2-way and 4-way)
+
+``` c
+  // Variants are also available for _s32_s16_s16 and _u32_u16_u16
+  __attribute__((arm_streaming_compatible))
+  svfloat32_t svdot[_f32_f16_f16](svfloat32_t zda, svfloat16_t zn, svfloat16_t zm);
```

Comment:
I wonder whether we should support `_n` versions of these functions too.  The corresponding 4-way functions allow things like:
```
svint32_t f(svint32_t x, svint8_t y) {
    return svdot(x, y, 1)
}
```
so it would be more consistent to also allow:
```
svint32_t f(svint32_t x, svint16_t y) {
    return svdot(x, y, 1)
}
```
The same thing applies to other arithmetic functions that have a single vector as their final operand.  E.g. perhaps the `[_single]` functions should also have an `[_n]` form that takes a scalar rather than a vector.  (The `[_n]` suffix would then replace the `[_single]` suffix, rather than add to it.)

Reply:
On second thoughts, the current `_n` versions of the `svdot` intrinsics rely on the fact that the third argument's type is predictable from the first two arguments' types.  In contrast, having three type suffixes allows the third argument's type to vary arbitrarily from the other two's, although in practice the most useful difference is probably going to be signed vs. unsigned.  So having `_n` forms of `svdot[_t0_t1_t2]` would mean using a scalar value to infer a type suffix for `t2`.  We've deliberately avoided inferring type suffixes from scalars in other contexts (e.g. `svdup`) because of the usual integer promotions.

So I suppose it's better to leave the `_n` form out after all.